tests/unit/asio_service_stream_test.cxx

```diff
@@ -29,57 +29,178 @@ using namespace raft_functional_common;
+            TestSuite::_msg("wait for receiving requests (up to %d ms)\n", timeout_ms);
+            req_ea_.wait_ms(1000);
+        }
+
+        // server
+        std::atomic<ulong> resp_log_index_;
+        std::atomic<bool> msg_mismatch_;
+        std::atomic<bool> reqs_out_of_order_;
+        EventAwaiter req_ea_;
+
+        // client
+        std::atomic<ulong> next_log_index_;
+        size_t sent_log_index_ = 0;
+        size_t num_messages_sent_ = 0;
+        size_t error_req_count_ = 0;
+        EventAwaiter ea_;
+
+        // log
+        ptr<logger_wrapper> server_logger_;
+        ptr<logger_wrapper> client_logger_;
+    };
+
     class stream_msg_handler : public nuraft::msg_handler {
     public:
         stream_msg_handler(context* ctx,
                            const init_options& opt,
-                           ptr<logger_wrapper> log_wrapper)
+                           ptr<stream_statistic> stream_stat)
             : msg_handler(ctx, opt)
-            , my_log_wrapper_(log_wrapper)
-            , streamed_log_index(0)
-            , msg_mismatch(false)
+            , stream_stat_(stream_stat)
             {}
 
         ptr<resp_msg> process_req(req_msg& req, const req_ext_params& ext_params) {
             ptr<resp_msg> resp = cs_new<resp_msg>(state_->get_term(),
                                                   msg_type::append_entries_response,
                                                   id_,
                                                   req.get_src());
-            if (req.get_last_log_idx() == streamed_log_index) {
-                streamed_log_index++;
-                resp->accept(streamed_log_index.load());
+            if (req.get_last_log_idx() == stream_stat_->resp_log_index_) {
+                stream_stat_->resp_log_index_++;
+                resp->accept(stream_stat_->resp_log_index_.load());
                 ptr<buffer> buf = req.log_entries().at(0)->get_buf_ptr();
                 buf->pos(0);
                 std::string buf_str = buf->get_str();
                 if (buf_str != TEST_MSG) {
-                    SimpleLogger* ll = my_log_wrapper_->getLogger();
+                    SimpleLogger* ll = stream_stat_->server_logger_->getLogger();
                     _log_info(ll, "resp str: %s", buf_str.c_str());
-                    msg_mismatch.store(true);
+                    stream_stat_->msg_mismatch_.store(true);
                 }
             } else {
-                SimpleLogger* ll = my_log_wrapper_->getLogger();
+                SimpleLogger* ll = stream_stat_->server_logger_->getLogger();
                 _log_info(ll, "req log index not match, req: %ld, current: %ld",
-                req.get_last_log_idx(), streamed_log_index.load());
+                req.get_last_log_idx(), stream_stat_->resp_log_index_.load());
+                stream_stat_->reqs_out_of_order_.store(true);
             }
+
+            stream_stat_->req_ea_.invoke();
```

Comment:
Do you intend invoking `req_ea_` at the first message? If so, put a comment at `req_ea_` definition.

Reply:
And I also see rare test failure here
```
[ .... ] server timeout test
   === TEST MESSAGE (BEGIN) ===
sending req: 1000/1000 (100.0%)
wait for receiving requests (up to 1000 ms)
wait for responses (up to 3000 ms)

        time: 2024-08-20 22:29:37.187010
      thread: ea83
          in: server_timeout_test()
          at: /home/junahn/work/raft/tests/unit/asio_service_stream_test.cxx:370
    value of: stat_ptr->waiting_for_responses()
    expected: true
      actual: false
[ FAIL ] server timeout test (3.1 s)
```
Haven't done detailed investigation, but probably be timing issue. You may need to either decrease `send_req` timeout (from 2s to 1s) or increase `wait_for_response` timeout (from 3s to 5s). Please take a look.